staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,1033 @@
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Attributes defines the set of attributes for this device.
+	// The name of each attribute must be unique in that set.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	// +listType=atomic
+	Attributes []DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,2,rep,name=attributes"`
+
+	// Capacities defines the set of capacities for this device.
+	// The name of each capacity must be unique in that set.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	// +listType=atomic
+	Capacities []DeviceCapacity `json:"capacities,omitempty" protobuf:"bytes,3,rep,name=capacities"`
+}
+
+// Limit for the sum of the number of entries in both ResourceSlices.
+const ResourceSliceMaxAttributesAndCapacitiesPerDevice = 32
+
+// DeviceAttribute is a combination of an attribute name and its value.
+// Exactly one value must be set.
+type DeviceAttribute struct {
+	// Name is a unique identifier for this attribute, which will be
+	// referenced when selecting devices.
+	//
+	// Attributes are defined either by the owner of the specific driver
+	// (usually the vendor) or by some 3rd party (e.g. the Kubernetes
+	// project). Because attributes are sometimes compared across devices,
+	// a given name is expected to mean the same thing and have the same
+	// type on all devices.
+	//
+	// Attribute names must be either a C identifier
+	// (e.g. "theName") or a DNS subdomain followed by a slash ("/")
+	// followed by a C identifier
+	// (e.g. "dra.example.com/theName"). Attributes whose name do not
+	// include the domain prefix are assumed to be part of the driver's
+	// domain. Attributes defined by 3rd parties must include the domain
+	// prefix.
+	//
+	// The maximum length for the DNS subdomain is 63 characters (same as
+	// for driver names) and the maximum length of the C identifier
+	// is 32.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// The Go field names below have a Value suffix to avoid a conflict between the
+	// field "String" and the corresponding method. That method is required.
+	// The Kubernetes API is defined without that suffix to keep it more natural.
+
+	// IntValue is a number.
+	//
+	// +optional
+	IntValue *int64 `json:"int,omitempty" protobuf:"varint,2,opt,name=int"`
+
+	// BoolValue is a true/false value.
+	//
+	// +optional
+	BoolValue *bool `json:"bool,omitempty" protobuf:"varint,3,opt,name=bool"`
+
+	// StringValue is a string. Must not be longer than 64 characters.
+	//
+	// +optional
+	StringValue *string `json:"string,omitempty" protobuf:"bytes,4,opt,name=string"`
+
+	// VersionValue is a semantic version according to semver.org spec 2.0.0.
+	// Must not be longer than 64 characters.
+	//
+	// +optional
+	VersionValue *string `json:"version,omitempty" protobuf:"bytes,5,opt,name=version"`
```

Comment:
@thockin: is it okay to limit the implementation of `matchAttribute` for a version to a literal string comparison?

In other words, `matchAttribute: driverVersion` only matches when two devices both use "1.0.0", but not when one device uses "1.0.0" and the other "1.0.00".

It is simpler to implement and more efficient. This needs to be documented, either here or under `matchAttribute`. I cannot imagine why vendors would publish the same version using a different string representation, so this shouldn't be a limitation in practice.

Reply:
Perhaps the "fix" is to document that version attributes are expected to be in canonical form - no leading zeros in numerical fields.  We should also document if they support semver's additional tags like `1.2.3+foo` or `5.6.78-something`, and whether it's expected to have a leading "v" or not.

Reply:
We implement semver 2.0. That's the spec. So no leading "v" (not allowed by semver 2.0) and yes, we support additional tags with the exact same sort order as defined by semver.

I am not sure whether there is a canonical form. I was assuming that there isn't and that both "1.0.0" and "1.0.00" are valid and are "equal" according to semver 2.0.

Reply:
Ack. Thanks for checking.

Reply:
I already says "according to semver.org spec 2.0.0", so the "xref the spec" part is covered?

I've updated the TODO remark in the implementation code.